crates/bitwarden/src/tool/generators/username.rs

```diff
@@ -0,0 +1,237 @@
+        include_number: Option<bool>,
+    },
+    /// Generates an email using your provider's subaddressing capabilities.
+    /// Note that not all providers support this functionality.
+    /// This will generate an address of the format `youremail+generated@domain.tld`
+    Subaddress {
+        /// The type of subaddress to add to the base email
+        r#type: AddressType,
+        /// The full email address to use as the base for the subaddress
+        email: String,
+    },
+    Catchall {
+        /// The type of username to use with the catchall email domain
+        r#type: AddressType,
+        /// The domain to use for the catchall email address
+        domain: String,
+    },
+    Forwarded {
+        /// The email forwarding service to use, see [ForwarderServiceType]
+        /// for instructions on how to configure each
+        service: ForwarderServiceType,
+        /// The website for which the email address is being generated
+        /// This is not used in all services, and is only used for display purposes
+        website: Option<String>,
+    },
+}
+
+#[derive(Serialize, Deserialize, Debug, JsonSchema)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct UsernameGeneratorRequest {
+    pub r#type: UsernameGeneratorType,
+}
+
+pub(super) async fn username(
+    input: UsernameGeneratorRequest,
+    http: &reqwest::Client,
+) -> Result<String> {
+    use rand::thread_rng;
+    use UsernameGeneratorType::*;
+
+    match input.r#type {
+        Word {
+            capitalize,
+            include_number,
+        } => {
+            let capitalize = capitalize.unwrap_or(true);
+            let include_number = include_number.unwrap_or(true);
+            Ok(username_word(&mut thread_rng(), capitalize, include_number))
+        }
+        Subaddress { r#type, email } => Ok(username_subaddress(&mut thread_rng(), r#type, email)),
+        Catchall { r#type, domain } => Ok(username_catchall(&mut thread_rng(), r#type, domain)),
+        Forwarded { service, website } => {
+            use crate::tool::generators::username_forwarders::*;
+            use ForwarderServiceType::*;
+            match service {
+                AddyIo {
+                    api_token,
+                    domain,
+                    base_url,
+                } => addyio::generate(http, api_token, domain, base_url, website).await,
+                DuckDuckGo { token } => duckduckgo::generate(http, token).await,
+                Firefox { api_token } => firefox::generate(http, api_token, website).await,
+                Fastmail { api_token } => fastmail::generate(http, api_token, website).await,
+                ForwardEmail { api_token, domain } => {
+                    forwardemail::generate(http, api_token, domain, website).await
+                }
+                SimpleLogin { api_key } => simplelogin::generate(http, api_key, website).await,
+            }
```

Comment:
Slightly clunky, would it make sense to extract this to a trait, or are we limited by the rust toolchain again? Would be nice to just do `service.generate()`

Reply:
Extracted to a generate function in ForwarderServiceType